zulip_bots/zulip_bots/lib.py

```diff
@@ -159,6 +161,48 @@ def state(self, default):
         yield new_state
         self.set_state(new_state)
 
+def setup_default_commands(bot_details, message_handler):
+    def def_about():
+        if bot_details['description'] == "":
+            return "**{}**".format(bot_details['name'])
+        return "**{}**: {}".format(bot_details['name'], bot_details['description'])
+
+    def def_help():
+        return ("\n".join("**{}** - {}".format(k, v[1])
+                          for k, v in defaults.items() if k) + "\n" +
+                "\n".join("**{}** - {}".format(k, v)
+                          for k, v in bot_details['commands'].items() if k))
+
+    def def_commands():
+        return "**Commands**: {} {}".format(
+            " ".join(k for k in defaults if k),
+            " ".join(k for k in bot_details['commands'] if k))
+    defaults = OrderedDict([  # Variable definition required for callbacks above
+        ('', (lambda: "Oops. Your message was empty.", "[BLANK MESSAGE NOT SHOWN]")),
+        ('about', (def_about, "The type and use of this bot")),
+        ('usage', ((lambda: message_handler.usage(), "Bot-provided usage text"))),
+        ('help', (lambda: "{}\n{}\n{}".format(def_about(), message_handler.usage(), def_help()),
+                  "This help text")),
+        ('commands', (def_commands, "A short list of supported commands"))
+    ])
+    return defaults
+
+def sync_botdetails_defaultcommands(bot_details, default_commands):
+    # Update default_commands from any changes in bot_details
+    if not bot_details['defaults']:  # Bot class will handle all commands
+        default_commands = {}
+    else:
+        if len(bot_details['commands']) == 0:  # No commands specified, so don't use this feature
+            del default_commands['commands']
+            del default_commands['help']
+        else:
+            for command in bot_details['commands']:  # Bot commands override defaults
+                if command in default_commands:
+                    del default_commands[command]
+    # Sync default_commands changes with bot_details
+    if len(default_commands) == 0:
+        bot_details['defaults'] = False
```

Comment:
This function would be much better as something that returned a new dictionary based on the inputs, rather than mutating the default_commands dict.  In addition to being a cleaner flow, I'm pretty sure the mutation means this code ends up being broken if multiple bots with different command sets run in the same process.